### PR TITLE
Controller refactor, documentation, capability queries 

### PR DESCRIPTION
**Pull request by gyrovorbis**
I noticed `controller.h` looked like it could use some formatting/cleanup attention... While I was in there, it dawned on me that perhaps this file is more confusing than it should be. The documentation wasn't as clear as it is in most places, and we had the internal condition struct from the RX frame declared publicly, with a long description of why it shouldn't be used and was replaced by `cont_state_t`. 

In reality, this structure doesn't look like it should be part of the API, as it seems fairly confusing to an end-user who just wants to see which API to use to grab input, and it doesn't look like there's anywhere else in the codebase that references the struct or any way for a client to grab ahold of the frame data to use it without doing bad things... So I made it private within the .c file... if everyone's cool with that. 

I also decided to create an anonymous union to allow me to add bitfields for each controller button state, providing a way for the end-user to check button states without having to do bitmasking (but also retaining backwards compatibility), which I think might be easier for them.

Finally, I really wanted to make the API and documentation make it extremely clear that the Dreamcast's controllers can support far more than is in the standard configuration... I wanted to make it very easy to check for something like the right analog or dpad or to see if there are extra buttons available, and make it part of the public API with documentation, to encourage people to take advantage of these kinds of devices in the future. 

I took the opportunity to also add capability "types" which can be used to query for a certain type of controller, based on its capabilities. I went ahead and did some research and experimentation with my own controllers to see what their reported capabilities were for populating these defines. I added:
- Standard Controller 
- Dual Analog Controller
- ASCII Pad
- Arcade Stick
- Twin Stick
- Racing Controller
- Maracas
- Fishing Rod

I just moved, haven't fully unpacked and have no idea where these are to test:
- Pop N Music Controller
- Densha De Go! Controller

Nothing that I have done has modified the existing API in any way... I'm just cleaning up, documenting, and expanding upon it. 

- moved cont_cond_t to .c file as a private structure, since nothing is using it publically, and it just confuses people
- added top-level "Controller" doxygen module for grouping everything related to the API
- added an anonymous union with button state bitfields to the cont_state_t structure, to simplify checking button states while also maintaining backwards compatibility
- fixed miscellaneous formatting issues
- reorganized controller.h to group things logically
- created "capability category" #defines for grouping together button capability masks into logical units
- "controller type" defines for providing capability presets for known controller types
- changed uintN types to uintN_t types
- documented everything ever (including controller callback's dirty little secret of executing within a disabled interrupts context)
- swapped a few sketchy C++ examples from directly casting their callback function to a manually-typed prototype (which didn't pefectly match, unsigned vs uint32_t, other sketchiness) to using the actual typedef within the header for casting. 

I have tested with all of the devices that I have available to me. I do not have my PopNMusic or Densha De Go! controllers readily available (just moved), so it will be quite awhile before I can get to them. I made a note in the code which will be generated in the Doxygen todo list. 

I also was generating the Doxygen the entire time I was working on this, and the documentation for the "Controller" module is now beautiful. :)

### CODE REVIEW

**Comment by gyrovorbis** [comment]
Screenshot from 2023-05-27 20-00-28
Here's a look at how the documentation is laid out and organized in Doxygen now for the Controller API.

File: kernel/arch/dreamcast/hardware/maple/controller.c [hunk 1]
```diff
@@ -11,22 +11,48 @@
...
+
 static cont_btn_callback_t btn_callback = NULL;
-static uint8 btn_callback_addr = 0;
-static uint32 btn_callback_btns = 0;
+static uint8_t btn_callback_addr = 0;
+static uint32_t btn_callback_btns = 0;
+
+/* Check whether the controller has EXACTLY the given capabilties. */
+int cont_is_type(const maple_device_t *cont, uint32_t type) {
+    return cont->info.function_data[CONT_FUNCTION_DATA_INDEX] == type;
```
Comment (QuzarDC) [inline kernel/arch/dreamcast/hardware/maple/controller.c:34]: Should these check for derefs, and -1 on bad input vs 0/1 for results?

File: kernel/arch/dreamcast/include/dc/maple.h [hunk 2]
```diff
@@ -192,6 +192,8 @@ typedef struct maple_frame {
 #endif
 } maple_frame_t;
 
+#define MAPLE_FRAME_WORD_BYTES  4  /**< \brief Bytes within a single word of a frame */
```
Comment (QuzarDC) [inline kernel/arch/dreamcast/include/dc/maple.h:195]: Is this something that should be its own define? It seems that a 'sizeof(uint32_t)' might be sufficient to be descriptive? The word size is also described in maple_frame_t->length. That's really minor tho. Seems we don't do that anywhere in /arch/dreamcast/. Just something to ponder.

File: kernel/arch/dreamcast/include/dc/maple/controller.h [hunk 3]
```diff
@@ -24,149 +27,466 @@
...
+
+                                ___________
+                               / |  __  |  \
+           L trigger -----|   /  | |  | |   \  |----- R trigger
+                         _|__/   | |__| |    \_|__
+                        |  _      \____/   _   _ |
+         Joystick   ----|-/ \             |X| |Y||
+                        | \_/              _   _ |
+                        |  _              |A| |B||
+                        |_| |_             ¯   ¯ |
```
Comment (QuzarDC) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:50]:
```suggestion
                        |  _      \____/          |
         Joystick   ----|-/ \               (Y)   |
                        | \_/            (X)   (B)|
                        |  _                (A)   |
                        |_| |_                    |
```

Comment (QuzarDC) [reply kernel/arch/dreamcast/include/dc/maple/controller.h:50]: Maybe? Putting them in a square seems really weird.

File: kernel/arch/dreamcast/include/dc/maple/controller.h [hunk 4]
```diff
@@ -24,149 +27,466 @@
...
+            uint32_t dpad2_down: 1;     /**< \brief Secondary Dpad Down button value. */
+            uint32_t dpad2_left: 1;     /**< \brief Secondary Dpad Left button value. */
+            uint32_t dpad2_right: 1;    /**< \brief Secondary Dpad Right button value. */
+            uint32_t: 16;
+        };
+    };
+
+    int ltrig;    /**< \brief Left trigger value (0-255). */
+    int rtrig;    /**< \brief Right trigger value (0-255). */
+    int joyx;     /**< \brief Main joystick x-axis value. (-128 - 127) */
```
Comment (QuzarDC) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:169]:
```suggestion
    int joyx;     /**< \brief Main joystick x-axis value (-128 - 127). */
```

File: kernel/arch/dreamcast/include/dc/maple/controller.h [hunk 5]
```diff
@@ -24,149 +27,466 @@
...
+    int joyy;     /**< \brief Main joystick y-axis value. */
+    int joy2x;    /**< \brief Secondary joystick x-axis value (if applicable). */
```
Comment (QuzarDC) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:171]: Aren't *all* the values in cont_state_t only 'if applicable'?

File: kernel/arch/dreamcast/include/dc/maple/controller.h [hunk 6]
```diff
@@ -24,149 +27,466 @@
...
     Functions of this type can be set with cont_btn_callback() to respond
     automatically to the specified set of buttons being pressed. This can be
     used, for instance, to implement the standard A+B+X+Y+Start method of ending
     the program running.
+
+    \warning
+    Your callback will be invoked within a context with interrupts disabled.
+    See cont_btn_callback for more information.
+
+    \param addr             Maple BUS address to poll for the button mask
```
Comment (QuzarDC) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:187]: Would it be worthwhile to note addr and btns with `maple_addr()` and the defgroup for btns?

Comment (QuzarDC) [reply kernel/arch/dreamcast/include/dc/maple/controller.h:187]: Oh, hah, you do it with the callback installer below. Excellent!

File: kernel/arch/dreamcast/include/dc/maple/controller.h [hunk 7]
```diff
@@ -24,149 +27,466 @@
...
+    \brief    API used to query for a controller's capabilities
+    \ingroup  controller
+    
+    The following API is used to query for the support of individual 
+    or groups of capabilities by a particular device.  
+*/
 
-/** \defgroup   controller_caps Controller capability bits.
+/** \defgroup controller_caps Capabilities
+    \brief    Bit masks used ot identify controller capabilities
```
Comment (QuzarDC) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:234]:
```suggestion
    \brief    Bit masks used to identify controller capabilities
```

File: kernel/arch/dreamcast/include/dc/maple/controller.h [hunk 8]
```diff
@@ -24,149 +27,466 @@
...
+    \ingroup  controller_query_caps
 
     These bits will be set in the function_data for the controller's deviceinfo
-    if the controller supports the corresponding feature.
+    if the controller supports the corresponding button/axis capability.
+
+    @{
+*/
+#define CONT_CAPABILITY_C               (1<<24)     /**< \brief C button capability mask. */
```
Comment (QuzarDC) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:242]: Might be worthwhile to note that the strange ordering is to match the above controller_inputs defgroup

File: kernel/arch/dreamcast/include/dc/maple/controller.h [hunk 9]
```diff
@@ -24,149 +27,466 @@
...
+                                             CONT_CAPABILITIES_EXTENDED_BUTTONS | \
+                                             CONT_CAPABILITIES_DUAL_ANALOG)
+
+/** \brief Fishing rod controller type */
+#define CONT_TYPE_FISHING_ROD               (CONT_CAPABILITIES_STANDARD_BUTTONS | \
+                                             CONT_CAPABILITIES_DPAD | \
+                                             CONT_CAPABILITIES_TRIGGERS | \
+                                             CONT_CAPABILITIES_DUAL_ANALOG)
+
+/** \todo Unknown Pop'n'Music controller type */
```
Comment (Tchan0) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:450]: Pop'n controller has A-B-C, X-Y, DPAD, and start, so:

/** \brief Pop'n'Music controller type */
#define CONT_TYPE_POP_N_MUSIC               (CONT_CAPABILITIES_STANDARD_BUTTONS | \
                                             CONT_CAPABILITY_C | \
                                             CONT_CAPABILITIES_DPAD)

File: kernel/arch/dreamcast/include/dc/maple/controller.h [hunk 10]
```diff
@@ -24,149 +27,466 @@
...
+#define CONT_TYPE_POP_N_MUSIC
+
+/** \todo Unknown de Go! controller type */
+#define CONT_TYPE_DENSHA_DE_GO
```
Comment (Tchan0) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:454]: Densha de go has A-B-C-D, X-Y-Z, DPAD and  START, so:

/** \brief Densha de Go! controller type */
#define CONT_TYPE_DENSHA_DE_GO              (CONT_CAPABILITIES_STANDARD_BUTTONS | \
                                             CONT_CAPABILITIES_EXTENDED_BUTTONS | \
                                             CONT_CAPABILITY_D | \
                                             CONT_CAPABILITIES_DPAD)

File: kernel/arch/dreamcast/include/dc/maple/controller.h [hunk 11]
```diff
@@ -24,149 +27,466 @@
...
 /** @} */
```
Comment (Tchan0) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:455]: ASCII Mission Stick:
- A-B-X-Y-START buttons -> CONT_CAPABILITIES_STANDARD_BUTTONS
- When pressing the DPAD, both DPAD 1 & 2 are received -> CONT_CAPABILITIES_DUAL_DPAD (although they will always have identical values)
- Left mouse L & R buttons = LEFT & RIGTH TRIGGER -> CONT_CAPABILITIES_TRIGGERS
- Right Mouse:
   * stick = JOYSTICK XY -> CONT_CAPABILITIES_ANALOG
   * the 3 buttons of the right mouse can be dynamically assigned to nothing/A/B/X/Y

so:

/** \brief ASCII Mission Stick controller type */
#define CONT_TYPE_ASCII_MISSION_STICK       (CONT_CAPABILITIES_STANDARD_BUTTONS | \
                                             CONT_CAPABILITIES_DUAL_DPAD | \
                                             CONT_CAPABILITIES_TRIGGERS | \
                                             CONT_CAPABILITIES_ANALOG)

Comment (Tchan0) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:455]: /** \brief Konami Dance Dance Revolution controller type */
#define CONT_TYPE_KONAMI_DDR                (CONT_CAPABILITY_A | \
                                             CONT_CAPABILITY_B | \
                                             CONT_CAPABILITY_START | \
                                             CONT_CAPABILITIES_DPAD)

Comment (gyrovorbis) [reply kernel/arch/dreamcast/include/dc/maple/controller.h:455]: omg thank you, @Tchan0 !!!!

File: kernel/arch/dreamcast/include/dc/maple/controller.h (hunk 5 above)
Comment (gyrovorbis) [reply kernel/arch/dreamcast/include/dc/maple/controller.h:171]: ...yes, lol. Good point.

File: kernel/arch/dreamcast/hardware/maple/controller.c (hunk 1 above)
Comment (gyrovorbis) [reply kernel/arch/dreamcast/hardware/maple/controller.c]: Added!

File: kernel/arch/dreamcast/include/dc/maple/controller.h [hunk 12]
```diff
@@ -1,20 +1,23 @@
 /* KallistiOS ##version##
 
    dc/maple/controller.h
-   (C)2000-2002 Jordan DeLong, Megan Potter
+   (C) 2000-2002 Jordan DeLong, Megan Potter
+   (C) 2023 Falco Girgis
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:5]:
```suggestion
   Copyright (C) 2000-2002 Jordan DeLong
   Copyright (C) 2000-2002 Megan Potter
   Copyright (C) 2023 Falco Girgis
```

File: kernel/arch/dreamcast/include/dc/maple/controller.h [hunk 13]
```diff
@@ -24,152 +27,494 @@
...
+                         \         /__\          /
+                          \    _____|_______    /
+                           \  /     |       \  /
+                            \/      |        \/
+                               Start button
+
+   You can grab a pointer to a connected controller by 
+   using the following:
+
+        maple_device_t* device = maple_enum_type(N, MAPLE_FUNC_CONTROLLER);
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:62]:
```suggestion
        maple_device_t *device = maple_enum_type(N, MAPLE_FUNC_CONTROLLER);
```

File: kernel/arch/dreamcast/include/dc/maple/controller.h [hunk 14]
```diff
@@ -24,152 +27,494 @@
...
-typedef struct {
-    uint16 buttons;         /* buttons bitfield */
-    uint8 rtrig;            /* right trigger */
-    uint8 ltrig;            /* left trigger */
-    uint8 joyx;         /* joystick X */
-    uint8 joyy;         /* joystick Y */
-    uint8 joy2x;            /* second joystick X */
-    uint8 joy2y;            /* second joystick Y */
-} cont_cond_t;
+        cont_state_t* state = (cont_state_t*)maple_dev_status(device);
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:80]:
```suggestion
        cont_state_t *state = (cont_state_t *)maple_dev_status(device);
```

**Review by ljsebald — CHANGES_REQUESTED** [review]
Minor cleanups... Otherwise, looks good.

**Comment by gyrovorbis** [comment]
Did one last sanity check to make sure nothing broke... everything is working!

**Review by ljsebald — APPROVED** [review]

**Pull request merged**